src/coreclr/debug/daccess/cdac.cpp

```diff
@@ -0,0 +1,79 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "cdac.h"
+#include <sospriv.h>
+#include <sstring.h>
+#include "dbgutil.h"
+
+#define CDAC_LIB_NAME MAKEDLLNAME_W(W("cdacreader"))
+
+namespace
+{
+    bool TryLoadCDACLibrary(HMODULE *phCDAC)
+    {
+        // Load cdacreader from next to DAC binary
+        PathString path;
+        if (FAILED(GetClrModuleDirectory(path)))
+            return false;
+
+        path.Append(CDAC_LIB_NAME);
+        *phCDAC = CLRLoadLibrary(path.GetUnicode());
+        if (*phCDAC == NULL)
+            return false;
+
+        return true;
+    }
+
+    int ReadFromTargetCallback(uint64_t addr, uint8_t* dest, uint32_t count, void* context)
+    {
+        CDAC* cdac = reinterpret_cast<CDAC*>(context);
+        return cdac->ReadFromTarget(addr, dest, count);
+    }
+}
+
+CDAC* CDAC::Create(uint64_t descriptorAddr, ICorDebugDataTarget* target)
+{
+    HMODULE cdacLib;
+    if (!TryLoadCDACLibrary(&cdacLib))
+        return nullptr;
+
+    CDAC *impl = new (nothrow) CDAC{cdacLib, descriptorAddr, target};
+    return impl;
+}
+
+CDAC::CDAC(HMODULE module, uint64_t descriptorAddr, ICorDebugDataTarget* target)
+    : m_module(module)
+    , m_target{target}
+{
+    decltype(&cdac_reader_init) init = reinterpret_cast<decltype(&cdac_reader_init)>(::GetProcAddress(m_module, "cdac_reader_init"));
+    decltype(&cdac_reader_get_sos_interface) getSosInterface = reinterpret_cast<decltype(&cdac_reader_get_sos_interface)>(::GetProcAddress(m_module, "cdac_reader_get_sos_interface"));
+    m_free = reinterpret_cast<decltype(&cdac_reader_free)>(::GetProcAddress(m_module, "cdac_reader_free"));
+    _ASSERTE(init != nullptr && getSosInterface != nullptr && m_free != nullptr);
+
+    init(descriptorAddr, &ReadFromTargetCallback, this, &m_cdac_handle);
+    getSosInterface(m_cdac_handle, &m_sos);
+}
+
+CDAC::~CDAC()
```

Comment:
What are the deallocation semantics here? I see a `new` so do we expect consumers to know to call `delete`? Is it possible we want to just return an instance and rely on stack and `move` semantics to handle ownership and lifetime?